README.md
# Chappy Police Man

We listen daily to different events of citizen insecurity that unfortunately occur in our country and we know that we could be part of the statistics. Our authorities, despite doing their best to combat this situation, is not enough.
When these events occur, we seek prompt help from our police, however, our reporting system is manual, bureaucratic and cumbersome, this generates a loss of vital time to take action and in many cases, the report is not specified.
This is how Chappy is your friend bot, which seeks to eliminate all those barriers and be the bridge that allows us to simplify this search for help.

[Access the Messenger experience](https://www.messenger.com/t/111154927148468)

![Messenger Experience](public/smartchappyfront1.png)

## How we build it

After brainstorming, we finally opted for the idea of helping people connect with the police service quickly and easily through an interactive chat.
For that, we support ourselves with the following technologies:

1. Node.js: A JavaScript runtime built on Chrome's V8 JavaScript engine
2. Express.js: Fast, unopinionated, minimalist web framework for Node.js
3. i18n: Lightweight simple translation module with dynamic json storage
4. Pdfkit: PDF document generation library for Node and the browser that makes creating complex, multi-page, printable documents easy.
5. Original Coast Clothing: Is a Messenger bot for a fictional clothing brand created to showcase key features of the Messenger Platform. 
6. Facebook API: Set of definitions and protocols for building and integrating Facebook into your apps.
## Requirements

- **Facebook Page:** Will be used as the identity of your messaging experience. When people chat with your page. To create a new Page, visit https://www.facebook.com/pages/create.
- **Facebook Developer Account:** Required to create new apps, which are the core of any Facebook integration. You can create a new developer account by going to the [Facebook Developers website](https://developers.facebook.com/) and clicking the "Get Started" button.
- **Facebook App:** Contains the settings for your Messenger automation, including access tokens. To create a new app, visit your [app dashboard](https://developers.facebook.com/apps).


#### How to boostrap Chappy Police Man

1. Go to your app Basic Settings, [Find your app here](https://developers.facebook.com/apps)
2. Save the **App ID** number and the **App Secret**

#### Grant  Messenger access to Chappy Police Man

1. Go to your app Dashboard
2. Under Add Product find Messenger and click Set Up
3. Now you should be in the App Messenger Settings
4. Under Access Tokens, click on Edit Permissions
5. Select the desired page and allow Manage and access Page conversations in Messenger
6. Select the desired page and an access token should appear
7. Get the Page ID from the page access token by using the [Access Token Debugger](https://developers.facebook.com/tools/debug/accesstoken/)
8. In the section Built-In NLP, select your page and enable the toggle

# Installation

Clone this repository on your local machine:

```bash
$ git clone https://github.com/nickthewitcher/chappy-police-man.git
$ cd chappy-police-man
```

You will need:

- [Node](https://nodejs.org/en/) 12.x or higher
- [Azure](https://azure.microsoft.com/) or remote server like [Heroku](https://www.heroku.com/) 

# Usage

## Using Heroku
#### 1. Install the Heroku CLI

Download and install the [Heroku CLI](https://devcenter.heroku.com/articles/heroku-cli)

#### 2. Create an app from the CLI

```bash
git init
heroku apps:create
# Creating app... done, ⬢ mystic-wind-83
# Created http://mystic-wind-83.herokuapp.com/ | git@heroku.com:mystic-wind-83.git
```

#### 3. Deploy the code
```bash
git add .
git commit -m "My first commit"
git push heroku master
```

#### 4. Set your environment variables
  In your Heroku App Dashboard [https://dashboard.heroku.com/apps/mystic-wind-83](https://dashboard.heroku.com/apps/mystic-wind-83) set up the config vars following the comments in the file ```.sample.env```

#### 5. Configure your webhook subscription and set the Messenger profile
  You should now be able to access the application. Use the ```VERIFY_TOKEN``` that you created as config vars and call the **/profile** endpoint like so:

  ```
  http://mystic-wind-83.herokuapp.com/profile?mode=all&verify_token=verify-token
  ```

#### 6. Test that your app setup is successful

  Send a message to your Page from Facebook or in Messenger, if your webhook receives an event, you have fully set up Chappy Police Man!

## License
Chappy Police Man is based in the Sample Messenger App Original Coast Clothing which is BSD licensed, as found in the LICENSE file.

See the [CONTRIBUTING](CONTRIBUTING.md) file for how to help out.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.0] - 2020-03-13
### Forked from the Original Coast Clothing
- First version of Chappy Police Man launched.


demo-table.js
"use strict";

const fs = require("fs");
const PDFDocument = require("./pdfkit-tables");
const i18n = require("./i18n.config");
//const User = require("./user");

//let user = new User("001");
module.exports = class DemoTable {
  constructor(user) {
    this.user = user;
  }
  /*
  function savePdfToFile (doc) {
    return new Promise((resolve, reject) => {

      // To determine when the PDF has finished being written successfully
      // we need to confirm the following 2 conditions:
      //
      //   1. The write stream has been closed
      //   2. PDFDocument.end() was called syncronously without an error being thrown

      let pendingStepCount = 2;

      const stepFinished = () => {
        if (--pendingStepCount == 0) {
          resolve();
        }
      };

      const writeStream = fs.createWriteStream(fileName);
      writeStream.on('close', stepFinished);
      doc.pipe(writeStream);

      doc.end();

      stepFinished();
    });
  }*/

  createPDF() {
    return new Promise((resolve, reject) => {
      //   1. The write stream has been closed
      //   2. PDFDocument.end() was called syncronously without an error being thrown

      const doc = new PDFDocument();

      let pendingStepCount = 2;

      const stepFinished = () => {
        console.log("llamada finished");
        if (--pendingStepCount == 0) {
          resolve("sucess");
          console.log("Demo-table success");
        }
      };

      if (pendingStepCount === 2) {
        console.log("Iniciando 2");
      } else if (pendingStepCount === 1) {
        console.log("Iniciando 1");
      } else if (pendingStepCount === 0) {
        console.log("Terminando 0");
      }

      const writeStream = fs.createWriteStream(
        __dirname + "/public/" + this.user.idDocument + ".pdf"
      );

      doc.pipe(writeStream);
      console.log("Empezando pipe");

      /*
      doc.pipe(
      fs.createWriteStream(
        __dirname + "/public/" + this.user.idDocument + ".pdf"
      )
    );*/

      doc.text(
        "\n" + i18n.__("report.idreport_en") + this.user.idreport,
        {
          align: "right"
        },
        70
      );

      doc.text(i18n.__("titles.title_en") + "\n\n\n\n\n", 130, 70);
      console.log("LLamando a doc.text 91");

      doc.image("./public/logo_chappy_police.png", 70, 40, {
        fit: [50, 50],
        align: "center",
        valign: "center"
      });

      const table0 = {
        headers: [i18n.__("titles.titl_per_en"), " "],
        rows: [
          [
            i18n.__("report.fullname_en"),
            this.user.firstName + " " + this.user.lastName
          ],
          [i18n.__("report.iddoc_en"), this.user.legalDni],
          [i18n.__("report.birthday_en"), this.user.legalBirthday],
          [i18n.__("report.telephone_en"), this.user.cellphone],
          [i18n.__("report.email_en"), this.user.legalEmail],
          [i18n.__("report.address_en"), this.user.legaladdress]
        ]
      };

      let reportString = "";
      //choose
      switch (this.user.typeOfReport) {
        case "yes_1":
          reportString = i18n.__("menu.complaints_1");
          break;
        case "yes_2":
          reportString = i18n.__("menu.complaints_2");
          break;
        case "yes_3":
          reportString = i18n.__("menu.complaints_3");
          break;
        case "yes_4":
          reportString = i18n.__("menu.complaints_4");
          break;
      }

      const table1 = {
        headers: [i18n.__("titles.title_com_en"), " "],
        rows: [
          [i18n.__("report.type_en"), reportString],
          [i18n.__("report.datetime_en"), this.user.dateOfFact],
          [i18n.__("report.place_en"), this.user.addressFact],
          [i18n.__("report.details_en"), this.user.howFact]
          //,        [i18n.__("report.evidence_en"), this.user.evidenceUrl]
        ]
      };

      doc.table(table0, table1, {
        prepareHeader: () => doc.font("Helvetica-Bold").fontsize(12),
        // eslint-disable-next-line no-unused-vars
        prepareRow: (row, i) => doc.font("Helvetica").fontSize(10)
      });

      doc.moveDown().table(table1, 70, 360);

      doc.moveDown();
      doc.text(
        "____________________________\n" + i18n.__("report.sign_authority"),
        350,
        670
      );
      console.log("LLamando a doc.text 139");

      doc.text(
        "____________________________\n" +
          i18n.__("report.sign_complainant") +
          "\n" +
          this.user.firstName +
          " " +
          this.user.lastName,
        70,
        670,
        {
          align: "left"
        }
      );

      doc.end();
      console.log("LLamando a doc.end");
      stepFinished();

      writeStream.on("finish", stepFinished);
      writeStream.on("error", e => {
        console.log("Error:");
        console.log(e);
        console.log("----------");
        reject(e);
      });
    });
  }
};


services/user.js
/**
 * Copyright 2019-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * Messenger For Original Coast Clothing
 * https://developers.facebook.com/docs/messenger-platform/getting-started/sample-apps/original-coast-clothing
 */

"use strict";
const { v4: uuidv4 } = require("uuid");
let random = uuidv4();

module.exports = class User {
  constructor(psid) {
    this.psid = psid;
    this.firstName = "";
    this.lastName = "";
    this.locale = "";
    this.timezone = "";
    this.gender = "neutral";
    this.state = "select_input";
    this.typeOfReport = "";
    this.complaintType = "";
    this.legalName = "";
    this.legalDni = "";
    this.legalBirthday = "";
    this.cellphone = "";
    this.legalEmail = "";
    this.legaladdress = "";
    this.photoUrl = "";
    this.dateOfFact = "";
    this.addressFact = "";
    this.howFact = "";
    this.detailFact = "";
    this.evidenceUrl = "";
    this.idreport = random.substring(30, 40);
    this.idDocument = random;
  }
  setProfile(profile) {
    this.firstName = profile.firstName;
    this.lastName = profile.lastName;
    this.locale = profile.locale;
    this.timezone = profile.timezone;
    if (profile.gender) {
      this.gender = profile.gender;
    }
  }
};


services/config.js
/**
 * Copyright 2019-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * Messenger For Original Coast Clothing
 * https://developers.facebook.com/docs/messenger-platform/getting-started/sample-apps/original-coast-clothing
 */

"use strict";

// Use dotenv to read .env vars into Node
require("dotenv").config();

// Required environment variables
const ENV_VARS = [
  "PAGE_ID",
  "APP_ID",
  "PAGE_ACCESS_TOKEN",
  "APP_SECRET",
  "VERIFY_TOKEN",
  "APP_URL",
  "SHOP_URL"
];

module.exports = {
  // Messenger Platform API
  mPlatformDomain: "https://graph.facebook.com",
  mPlatformVersion: "v3.2",

  // Page and Application information
  pageId: process.env.PAGE_ID,
  appId: process.env.APP_ID,
  pageAccesToken: process.env.PAGE_ACCESS_TOKEN,
  appSecret: process.env.APP_SECRET,
  verifyToken: process.env.VERIFY_TOKEN,

  // URL of your app domain
  appUrl: process.env.APP_URL,

  // URL of your website
  shopUrl: process.env.SHOP_URL,

  // Persona IDs
  personas: {},

  // Preferred port (default to 3000)
  port: process.env.PORT || 3000,

  get mPlatfom() {
    return this.mPlatformDomain + "/" + this.mPlatformVersion;
  },

  // URL of your webhook endpoint
  get webhookUrl() {
    return this.appUrl + "/webhook";
  },

  get botUrl() {
    return this.appUrl;
  },

  get newPersonas() {
    return [
      {
        name: "Jorge",
        picture: `${this.appUrl}/personas/sales.jpg`
      },
      {
        name: "Laura",
        picture: `${this.appUrl}/personas/billing.jpg`
      },
      {
        name: "Riandy",
        picture: `${this.appUrl}/personas/order.jpg`
      },
      {
        name: "Daniel",
        picture: `${this.appUrl}/personas/care.jpg`
      }
    ];
  },

  pushPersona(persona) {
    this.personas[persona.name] = persona.id;
  },

  get personaSales() {
    let id = this.personas["Jorge"] || process.env.PERSONA_SALES;
    return {
      name: "Jorge",
      id: id
    };
  },

  get personaBilling() {
    let id = this.personas["Laura"] || process.env.PERSONA_BILLING;
    return {
      name: "Laura",
      id: id
    };
  },

  get personaOrder() {
    let id = this.personas["Riandy"] || process.env.PERSONA_ORDER;
    return {
      name: "Riandy",
      id: id
    };
  },

  get personaCare() {
    let id = this.personas["Daniel"] || process.env.PERSONA_CARE;
    return {
      name: "Daniel",
      id: id
    };
  },

  get whitelistedDomains() {
    return [this.appUrl, this.shopUrl];
  },

  checkEnvVariables: function() {
    ENV_VARS.forEach(function(key) {
      if (!process.env[key]) {
        console.log("WARNING: Missing the environment variable " + key);
      } else {
        // Check that urls use https
        if (["APP_URL", "SHOP_URL"].includes(key)) {
          const url = process.env[key];
          if (!url.startsWith("https://")) {
            console.log(
              "WARNING: Your " + key + ' does not begin with "https://"'
            );
          }
        }
      }
    });
  }
};


app.js
/**
 * Copyright 2019-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * Messenger For Original Coast Clothing
 * https://developers.facebook.com/docs/messenger-platform/getting-started/sample-apps/original-coast-clothing
 */

"use strict";

// Imports dependencies and set up http server
const express = require("express"),
  { urlencoded, json } = require("body-parser"),
  crypto = require("crypto"),
  path = require("path"),
  Receive = require("./services/receive"),
  DemoTable = require("./demo-table"),
  GraphAPi = require("./services/graph-api"),
  User = require("./services/user"),
  config = require("./services/config"),
  i18n = require("./i18n.config"),
  app = express();
const { v4: uuidv4 } = require("uuid");
var users = {};
// Parse application/x-www-form-urlencoded
app.use(
  urlencoded({
    extended: true
  })
);

//Test test
// Parse application/json. Verify that callback came from Facebook
app.use(json({ verify: verifyRequestSignature }));

// Serving static files in Express
app.use(express.static(path.join(path.resolve(), "public")));

// Set template engine in Express
app.set("view engine", "ejs");

// Respond with index file when a GET request is made to the homepage
app.get("/", function(_req, res) {
  res.render("index");
});

// Adds support for GET requests to our webhook
app.get("/webhook", (req, res) => {
  // Parse the query params
  let mode = req.query["hub.mode"];
  let token = req.query["hub.verify_token"];
  let challenge = req.query["hub.challenge"];

  // Checks if a token and mode is in the query string of the request
  if (mode && token) {
    // Checks the mode and token sent is correct
    if (mode === "subscribe" && token === config.verifyToken) {
      // Responds with the challenge token from the request
      console.log("WEBHOOK_VERIFIED");
      res.status(200).send(challenge);
    } else {
      // Responds with '403 Forbidden' if verify tokens do not match
      res.sendStatus(403);
    }
  }
});

// Creates the endpoint for your webhook
app.post("/webhook", (req, res) => {
  let body = req.body;
  //test
  // Checks if this is an event from a page subscription
  if (body.object === "page") {
    // Returns a '200 OK' response to all requests
    res.status(200).send("EVENT_RECEIVED");

    // Iterates over each entry - there may be multiple if batched
    body.entry.forEach(function(entry) {
      // Get the webhook event. entry.messaging is an array, but
      // will only ever contain one event, so we get index 0
      if ("changes" in entry) {
        console.log("Changes Happend");
        // Handle Page Changes event
        let receiveMessage = new Receive();
        if (entry.changes[0].field === "feed") {
          let change = entry.changes[0].value;
          switch (change.item) {
            case "post":
              return receiveMessage.handlePrivateReply(
                "post_id",
                change.post_id
              );
              break;
            case "comment":
              return receiveMessage.handlePrivateReply(
                "commentgity _id",
                change.comment_id
              );
              break;
            default:
              console.log("Unsupported feed change type.");
              return;
          }
        }
      }

      // Gets the body of the webhook event
      let webhook_event = entry.messaging[0];
      //console.log("Gets the body of the webhook event ");
      //console.log(webhook_event);

      // Discard uninteresting events
      if ("read" in webhook_event) {
        //console.log("Got a read event xd ");
        return;
      }

      if ("delivery" in webhook_event) {
        //console.log("Got a delivery event xd");
        return;
      }

      // Get the sender PSID
      let senderPsid = webhook_event.sender.id;

      if (!(senderPsid in users)) {
        let user = new User(senderPsid);
        console.log("Entro a Crear User");
        GraphAPi.getUserProfile(senderPsid)
          .then(userProfile => {
            user.setProfile(userProfile);
          })
          .catch(error => {
            // The profile is unavailable
            console.log("Profile is unavailable:", error);
          })
          .finally(() => {
            users[senderPsid] = user;
            i18n.setLocale(user.locale);
            console.log(
              "New Profile PSID:",
              senderPsid,
              "with locale:",
              i18n.getLocale()
            );
            console.log("Full user");
            console.log(users[senderPsid]);
            let receiveMessage = new Receive(users[senderPsid], webhook_event);
            return receiveMessage.handleMessage(user.state);
          });
      } else {
        console.log("Entro a Asignar User");
        i18n.setLocale(users[senderPsid].locale);
        console.log(
          "Profile already exists PSID:",
          senderPsid,
          "with locale:",
          i18n.getLocale()
        );
        console.log("Full user");
        console.log(users[senderPsid]);
        let receiveMessage = new Receive(users[senderPsid], webhook_event);
        try {
          if (webhook_event.postback) {
            if (webhook_event.postback.title) {
              if (webhook_event.postback.title === "Get Started") {
                console.log("Inicializamos en postback get started");
                users[senderPsid].state = "getstarted";
                users[senderPsid].typeOfReport = "";
                users[senderPsid].legalName = "";
                users[senderPsid].legalDni = "";
                users[senderPsid].legalBirthday = "";
                users[senderPsid].cellphone = "";
                users[senderPsid].legalEmail = "";
                users[senderPsid].legaladdress = "";
                users[senderPsid].photoUrl = "";
                users[senderPsid].dateOfFact = "";
                users[senderPsid].addressFact = "";
                users[senderPsid].howFact = "";
                users[senderPsid].detailFact = "";
                users[senderPsid].evidenceUrl = "";
                let random = uuidv4();
                console.log("Generado nuevo random");
                console.log(random);
                users[senderPsid].idreport = random.substring(30, 40);
                users[senderPsid].idDocument = random;
              }
            }
          }
          //webhookEvent.message.quick_reply.payload
          if (webhook_event.message) {
            if (webhook_event.message.quick_reply) {
              if (webhook_event.message.quick_reply.payload) {
                if (
                  webhook_event.message.quick_reply.payload ===
                  "deny_confirmation"
                ) {
                  users[senderPsid].state = "nombre";
                  console.log("Deny confirmation app.js");
                } else if (
                  webhook_event.message.quick_reply.payload ===
                  "decline_evidence"
                ) {
                  users[senderPsid].state = "finish";
                  console.log("decline confirmation app.js");

                  console.log("PROCESO TERMINADO sin file ADJUNTO");
                  let pdfReport = new DemoTable(users[senderPsid]);

                  pdfReport.createPDF().then(successMessage => {
                    // succesMessage es lo que sea que pasamos en la función resolve(...) de arriba.
                    // No tiene por qué ser un string, pero si solo es un mensaje de éxito, probablemente lo sea.
                    console.log("¡Sí! " + successMessage);
                    console.log("Se creo el PDF");
                  });
                } else if (
                  webhook_event.message.quick_reply.payload ===
                  "accept_evidence"
                ) {
                  console.log("accept evidence  app.js");

                  users[senderPsid].state = "evidence_confirmed";
                }
              }
            }
          }
        } catch (error) {
          console.error(error);
          console.log(`An error has occured: '${error}'. We have been notified and \
        will fix the issue shortly!`);
        }
        console.log("Envia el mensaje a analizar si es un message or postback");
        console.log("-------------------------------------");
        console.log("Estado a entrar");
        console.log(users[senderPsid].state);
        let varResponse = receiveMessage.handleMessage(users[senderPsid].state);
        if (varResponse != null) {
          if (varResponse.length === 4) {
            if (varResponse[3].payload === "mygreetings") {
              console.log("My greetings");
              users[senderPsid].state = "getstarted";
              users[senderPsid].typeOfReport = "";
              users[senderPsid].legalName = "";
              users[senderPsid].legalDni = "";
              users[senderPsid].legalBirthday = "";
              users[senderPsid].cellphone = "";
              users[senderPsid].legalEmail = "";
              users[senderPsid].legaladdress = "";
              users[senderPsid].photoUrl = "";
              users[senderPsid].dateOfFact = "";
              users[senderPsid].addressFact = "";
              users[senderPsid].howFact = "";
              users[senderPsid].detailFact = "";
              users[senderPsid].evidenceUrl = "";
              let random = uuidv4();
              console.log("Generado nuevo random");
              console.log(random);
              users[senderPsid].idreport = random.substring(30, 40);
              users[senderPsid].idDocument = random;
            }
          } else if (varResponse.length === 2) {
            if (varResponse[1].payload === "mygreetings") {
              console.log("My greetings");
              users[senderPsid].state = "getstarted";
              users[senderPsid].typeOfReport = "";
              users[senderPsid].legalName = "";
              users[senderPsid].legalDni = "";
              users[senderPsid].legalBirthday = "";
              users[senderPsid].cellphone = "";
              users[senderPsid].legalEmail = "";
              users[senderPsid].legaladdress = "";
              users[senderPsid].photoUrl = "";
              users[senderPsid].dateOfFact = "";
              users[senderPsid].addressFact = "";
              users[senderPsid].howFact = "";
              users[senderPsid].detailFact = "";
              users[senderPsid].evidenceUrl = "";
              let random = uuidv4();
              console.log("Generado nuevo random");
              console.log(random);
              users[senderPsid].idreport = random.substring(30, 40);
              users[senderPsid].idDocument = random;
            }
          }
        }
        console.log("El response a retornar:");
        console.log(varResponse);
        if (
          users[senderPsid].state === "recomendation1" ||
          users[senderPsid].state === "recomendation2" ||
          users[senderPsid].state === "recomendation3" ||
          (users[senderPsid].state === "recomendation4" && varResponse != null)
        ) {
          users[senderPsid].howFact = varResponse[2].payload;
        }
        if (users[senderPsid].state === "getstarted" && varResponse != null) {
          users[senderPsid].state = "select_input";
        } else if (
          users[senderPsid].state === "select_input" &&
          varResponse != null
        ) {
          console.log(
            "complaintType a asignar: " + varResponse[1].complaintType
          );
          users[senderPsid].complaintType = varResponse[1].complaintType;
          users[senderPsid].state = "confirm_input";
        } else if (
          users[senderPsid].state === "confirm_input" &&
          varResponse != null
        ) {
          console.log(
            "Confirm input stado , a asignar: " + varResponse[1].payload
          );
          users[senderPsid].typeOfReport = varResponse[1].payload;
          users[senderPsid].state = "nombre";
        } else if (
          users[senderPsid].state === "nombre" &&
          varResponse != null
        ) {
          console.log("nombre estado a asignar: " + varResponse[1].payload);
          users[senderPsid].legalName = varResponse[1].payload;
          users[senderPsid].state = "dni";
        } else if (users[senderPsid].state === "dni" && varResponse != null) {
          users[senderPsid].legalDni = varResponse[1].payload;
          users[senderPsid].state = "birthday";
        } else if (
          users[senderPsid].state === "birthday" &&
          varResponse != null
        ) {
          users[senderPsid].legalBirthday = varResponse[1].payload;
          users[senderPsid].state = "cel";
        } else if (users[senderPsid].state === "cel" && varResponse != null) {
          users[senderPsid].cellphone = varResponse[1].payload;
          users[senderPsid].state = "email";
        } else if (users[senderPsid].state === "email" && varResponse != null) {
          users[senderPsid].state = "address";
          users[senderPsid].legalEmail = varResponse[1].payload;
        } else if (
          users[senderPsid].state === "address" &&
          varResponse != null
        ) {
          users[senderPsid].legaladdress = varResponse[1].payload;
          users[senderPsid].state = "photoquestion";
        } else if (
          users[senderPsid].state === "photoquestion" &&
          varResponse != null
        ) {
          users[senderPsid].state = "preparationquestion";
        } else if (
          users[senderPsid].state === "preparationquestion" &&
          varResponse != null
        ) {
          users[senderPsid].state = "when";
        } else if (users[senderPsid].state === "when" && varResponse != null) {
          users[senderPsid].dateOfFact = varResponse[1].payload;
          users[senderPsid].state = "where";
        } else if (users[senderPsid].state === "where" && varResponse != null) {
          users[senderPsid].addressFact = varResponse[2].payload;
          if (users[senderPsid].typeOfReport === "yes_1") {
            users[senderPsid].state = "recomendation1";
          }
          if (users[senderPsid].typeOfReport === "yes_2") {
            users[senderPsid].state = "recomendation2";
          }
          if (users[senderPsid].typeOfReport === "yes_3") {
            users[senderPsid].state = "recomendation3";
          }
          if (users[senderPsid].typeOfReport === "yes_4") {
            users[senderPsid].state = "recomendation4";
          }
        } else if (
          users[senderPsid].state === "evidence_confirmed" &&
          varResponse != null
        ) {
          users[senderPsid].state = "file_input";
        } else if (
          users[senderPsid].state === "file_input" &&
          varResponse != null
        ) {
          users[senderPsid].state = "finish";
          console.log("PROCESO TERMINADO con file ADJUNTO");
          let pdfReport = new DemoTable(users[senderPsid]);

          pdfReport.createPDF().then(successMessage => {
            // succesMessage es lo que sea que pasamos en la función resolve(...) de arriba.
            // No tiene por qué ser un string, pero si solo es un mensaje de éxito, probablemente lo sea.
            console.log("¡Sí! " + successMessage);
            console.log("Se creo el PDF");
          });
        }
        console.log("Estado cambiado");
        console.log("-------------------------------------");
        console.log(users[senderPsid].state);
        return varResponse;
      }
    });
  } else {
    // Returns a '404 Not Found' if event is not from a page subscription
    console.log("Not found 404");
    res.sendStatus(404);
  }
});

// Set up your App's Messenger Profile
app.get("/profile", (req, res) => {
  let token = req.query["verify_token"];
  let mode = req.query["mode"];

  if (!config.webhookUrl.startsWith("https://")) {
    res.status(200).send("ERROR - Need a proper API_URL in the .env file");
  }
  var Profile = require("./services/profile.js");
  Profile = new Profile();
  // Checks if a token and mode is in the query string of the request
  if (mode && token) {
    if (token === config.verifyToken) {
      if (mode == "webhook" || mode == "all") {
        Profile.setWebhook();
        res.write(
          `<p>Set app ${config.appId} call to ${config.webhookUrl}</p>`
        );
      }
      if (mode == "profile" || mode == "all") {
        Profile.setThread();
        res.write(`<p>Set Messenger Profile of Page ${config.pageId}</p>`);
      }
      if (mode == "personas" || mode == "all") {
        Profile.setPersonas();
        res.write(`<p>Set Personas for ${config.appId}</p>`);
        res.write(
          "<p>To persist the personas, add the following variables \
          to your environment variables:</p>"
        );
        res.write("<ul>");
        res.write(`<li>PERSONA_BILLING = ${config.personaBilling.id}</li>`);
        res.write(`<li>PERSONA_CARE = ${config.personaCare.id}</li>`);
        res.write(`<li>PERSONA_ORDER = ${config.personaOrder.id}</li>`);
        res.write(`<li>PERSONA_SALES = ${config.personaSales.id}</li>`);
        res.write("</ul>");
      }
      if (mode == "nlp" || mode == "all") {
        GraphAPi.callNLPConfigsAPI();
        res.write(`<p>Enable Built-in NLP for Page ${config.pageId}</p>`);
      }
      if (mode == "domains" || mode == "all") {
        Profile.setWhitelistedDomains();
        res.write(`<p>Whitelisting domains: ${config.whitelistedDomains}</p>`);
      }
      if (mode == "private-reply") {
        Profile.setPageFeedWebhook();
        res.write(`<p>Set Page Feed Webhook for Private Replies.</p>`);
      }
      res.status(200).end();
    } else {
      // Responds with '403 Forbidden' if verify tokens do not match
      res.sendStatus(403);
    }
  } else {
    // Returns a '404 Not Found' if mode or token are missing
    res.sendStatus(404);
  }
});

// Verify that the callback came from Facebook.
function verifyRequestSignature(req, res, buf) {
  var signature = req.headers["x-hub-signature"];

  if (!signature) {
    console.log("Couldn't validate the signature.");
  } else {
    var elements = signature.split("=");
    var signatureHash = elements[1];
    var expectedHash = crypto
      .createHmac("sha1", config.appSecret)
      .update(buf)
      .digest("hex");
    if (signatureHash != expectedHash) {
      throw new Error("Couldn't validate the request signature.");
    }
  }
}

// Check if all environment variables are set
config.checkEnvVariables();

// listen for requests :)
var listener = app.listen(config.port, function() {
  console.log("Your app is listening on port " + listener.address().port);

  if (
    Object.keys(config.personas).length == 0 &&
    config.appUrl &&
    config.verifyToken
  ) {
    console.log(
      "Is this the first time running?\n" +
        "Make sure to set the both the Messenger profile, persona " +
        "and webhook by visiting:\n" +
        config.appUrl +
        "/profile?mode=all&verify_token=" +
        config.verifyToken
    );
  }

  if (config.pageId) {
    console.log("Test your app by messaging:");
    console.log("https://m.me/" + config.pageId);
  }
});
